rclpy/rclpy/executors.py

```diff
@@ -745,24 +745,45 @@ class SingleThreadedExecutor(Executor):
     def __init__(self, *, context: Optional[Context] = None) -> None:
         super().__init__(context=context)
 
-    def spin_once(self, timeout_sec: Optional[float] = None) -> None:
+        self._futures = []
+
+    def _spin_once_impl(
+        self,
+        timeout_sec: Optional[float] = None,
+        wait_condition: Callable[[], bool] = lambda: False
+    ) -> None:
         try:
-            handler, entity, node = self.wait_for_ready_callbacks(timeout_sec=timeout_sec)
+            handler, entity, node = self.wait_for_ready_callbacks(
+                timeout_sec, None, wait_condition)
         except ShutdownException:
             pass
         except TimeoutException:
             pass
+        except ConditionReachedException:
+            pass
```

Comment:
just a comment to make sure if i understand correctly.

if the condition (future) is reached, leads to this exception to pass. this is the one of the case that `wait_condition` is already completed. (directly related to https://github.com/ros2/rclpy/issues/989)

Reply:
Yes, exactly.  From what I can tell looking at the code, the reason it is done this way is because we had to call the `wait_condition` to discover whether it was completed.  If it *was* completed, then presumably it changed some state already (like setting the future to done, etc), and so we don't want to make that call again.